### ISSUE TITLE
Saving Estimators, loading from checkpoints and accessing placeholders.

### ISSUE THREAD

**Issue by zeyademam** [tensorflow/tensorflow#19041]
I posted this question here on stackoverflow (https://stackoverflow.com/questions/50011969/tensorflow-estimator-api-input-tensor-names) and someone else posted a similar question earlier (https://stackoverflow.com/questions/48335699/tensorflow-what-are-the-input-nodes-for-tf-estimator-models), yet no satisfying answers. 

When using an estimator in tensorflow and passing the inputs using `tf.estimator.inputs.numpy_input_function()`, I would like to find a way to access the input placeholders (one for features and one for labels).
I train the model using `tf.estimator.estimator.train()` and save the checkpoint files in `model_dir`. 
Later, I would like to load the model and variables using the following snippet:
`sess = tf.Session()`
`saver = tf.train.import_meta_graph(model_dir+'/model.ckpt-1.meta')`
`saver.restore(sess, tf.train.latest_checkpoint(model_dir))`
`predictions = graph.get_tensor_by_name('softmax_tensor:0')`

Then do `sess.run(predictions,....)`. But there's no way to pull out the inputs' placeholder for the `feed_dict` parameter of `sess.run()`.

Another related issue I have: I would also like to add placeholders to an estimator other than the features and labels placeholders. For example if I have dropout, I would like the dropout_rate or the `tf.layers.dropout()` function's `training` param to be placeholders. This way after loading the model from chkpt as above, I can turn off dropout. 

Have I written custom code Y
OS Platform and Distribution Ubuntu 18.04
TensorFlow installed from pip
TensorFlow version r1.8.0
Bazel version N/A
CUDA/cuDNN version 9.0/7.0.1
GPU model and memory GTX 1080TI
Exact command to reproduce N/A

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by rohan100jain** [comment]
Can you please elaborate how you use placeholders in the code here? Some code snippet on how you're setting those up and passing them into the numpy_input_fn?

**Comment by zeyademam** [comment]
I am not using place holders in setting up and training the estimator, this is done internally by the estimator object. Here is a code snippet on how I set up the estimator and train it.

```python
import numpy as np
import tensorflow as tf
from keras.datasets import cifar10

def model_fn(features, labels, mode, params):

    x = tf.reshape(features['x'], [-1, 32, 32, 3])
    x = tf.to_float(x)
    flatten = tf.layers.flatten(x)
    logits = tf.layers.dense(inputs=flatten,
                             units=10)


    predictions = {
        'classes': tf.argmax(input=logits, axis=1),
        'probabilities': tf.nn.softmax(logits, name='softmax_tensor')
    }

    # Calculate Loss (for both TRAIN and EVAL modes)
    l2_loss = tf.losses.get_regularization_losses()
    loss = tf.losses.sparse_softmax_cross_entropy(labels=labels, logits=logits)
    loss += tf.reduce_sum(l2_loss)
    # Calculate Accuracy (for both TRAIN and EVAL modes)
    accuracy = tf.metrics.accuracy(labels=labels,
                                   predictions=predictions['classes'],
                                   name='acc_op')
    metrics = {'accuracy': accuracy}

    if mode == tf.estimator.ModeKeys.TRAIN:
        optimizer = tf.train.AdamOptimizer()
        learning_rate = optimizer._lr
        train_op = optimizer.minimize(
            loss=loss,
            global_step=tf.train.get_global_step())


        return tf.estimator.EstimatorSpec(mode=mode,
                                          loss=loss,
                                          train_op=train_op)


    # Add evaluation metrics (for EVAL mode)
    if mode == tf.estimator.ModeKeys.EVAL:
        return tf.estimator.EstimatorSpec(mode=mode,
                                          loss=loss,
                                          eval_metric_ops=metrics)


if __name__== '__main__':
    # Load CIFAR10 data
    (x_train, y_train), (x_test, y_test) = cifar10.load_data()

    # Cast y arrays into int32
    y_train = np.array(y_train, dtype=np.int32)

    # Create the Estimator
    model_dir = '/home/zeyad/delete_me/'
    classifier = tf.estimator.Estimator(
        model_fn=model_fn,
        params=None,
        model_dir=model_dir)

    input_fn = tf.estimator.inputs.numpy_input_fn(x={'x':x_train},
                                                  y=y_train,
                                                  batch_size=1, shuffle=True)

    classifier.train(
        input_fn=input_fn,
        max_steps=5)

```

Now after the above saves the checkpoint files, I would like to load those checkpoint files using the following snippet, but I am not sure how to retrieve the input placeholder by name:

```python
import tensorflow as tf

model_dir = '/home/zeyad/delete_me/'
def attack():

    sess = tf.Session()
    saver = tf.train.import_meta_graph(model_dir+'/model.ckpt-1.meta')
    saver.restore(sess, tf.train.latest_checkpoint(model_dir))
    graph = tf.get_default_graph()

    inputs = graph.get_tensor_by_name()     # what do i put here?
```

If I were able to do that then i can implement things such as the adversarial attacks, disabling dropout, etc...

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 16 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 31 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 46 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by HoodKhizer** [comment]
I have run into the exact same problem, after training the network with the estimator api I would like to add placeholder input notes to graph in order to deploy it on edge and I just don't seem to be able to find a way to do this.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Arjuna197** [comment]
If you view your graph in tensorboard, you can pretty easily discern the name of the tensors you'll need at inference time.
Then, something like: 

```
saver.restore(sess, tf.train.latest_checkpoint(model_dir))
gd = sess.graph_def
inp, prediction = tf.import_graph_def(
  gd, return_elements = ['input_node:0', 'output_node/BiasAdd:0'])
result = sess.run(prediction, feed_dict={inp: my_prediction_inputs})
```

you can specify more than one input and output also. However, if your restored graph has batch_norm or dropout you may run into issues with this method. There is also probably a more elegant solution, but I just gave up and stopped using estimators in favor of having direct access to the session and layers and both training and inference time.

**Comment by zeyademam** [comment]
Thank you. I tried using tensorboard to do so for my [code snippet above](https://github.com/tensorflow/tensorflow/issues/19041#issuecomment-389611457), however, the names of the tensors are not easy to discern.  In that code snippet, the first computation that I define in the graph is a `reshape` op. Tensorboard shows the input to that op as another op called `random_shuffle_queue_DequeueUpTo` which is not a placeholder. And once again the inputs to that op is `enqueue_input/random_shuffle_queue` which is also an op and not a placeholder - this latter op has no inputs. I was not able to trace back to the original placeholder. I am not sure if I am missing something obvious. 

As for using the lower level api, I do so as well, however, when generating adversarial examples for instance, it is easier to download a trained network from the web and work with it. Having access to the input placeholder would make this feasible (i ended up re-purposing the PREDICT mode in the estimator to produce adversarial examples and it looks really whacky.)

**Comment by Arjuna197** [comment]
@zeyademam In the example I gave you should be able to choose arbitrary '<opname>:<output_index>' pairings as your "placeholder". This may be technically incorrect, but you may want to think of all node:output_index pairings in a loaded graph as placeholders for the purpose of picking inputs and outputs. Think of it like you're going to the store, starting from your house. The entry point isn't your house, its the first street you take when you walk out the door. Again, no promises on how accurate this metaphor is from a technical standpoint .

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 59 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 74 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by baregawi** [comment]
I have a general set of questions:
Why is saving and reloading models in Keras a few lines and every tutorial about how to save and reload tensorflow estimators is at least dozens of lines. Is there no way to treat a session, graph and a set of tensorflow variables/tensors as a single entity? Doesn't Keras already exist to do that? Many things only support tensorflow and not Keras so converting a Keras model to tensorflow estimators is useful at times (thank you for that feature!). But saving them seems to necessarily involve manually handling the session, graph and set of variables separately. Is there a way around this that I am not aware of?

**Comment by zeyademam** [comment]
I think loading estimators to keep training, evaluating, or perform predictions is fairly simple. 
My issue is with loading estimators in a lower level api to do more than what those three estimator ModeKeys allow. This is pretty difficult to do in my opinion. I gave up and used a workaround.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rohan100jain: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by vkaul11** [comment]
> bly a more elegant solution, but I just gave up and stopped using estimators in favor of having direct access to the session and layers and both training and inference time.

What do you use instead of estimators? Aren't those good for training? Also what is the change in tf 2.0 ?

**Comment by Saduf2019** [comment]
@zeyademam
Could you please try on latest stable version of tf and let us know if this is still an issue.Thanks!
